plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts

```diff
@@ -49,23 +47,92 @@ import {
 import { TeamMember as AdoTeamMember } from 'azure-devops-node-api/interfaces/common/VSSInterfaces';
 import { Logger } from 'winston';
 import { PolicyEvaluationRecord } from 'azure-devops-node-api/interfaces/PolicyInterfaces';
-import { WebApi } from 'azure-devops-node-api';
+import {
+  WebApi,
+  getHandlerFromToken,
+  getPersonalAccessTokenHandler,
+} from 'azure-devops-node-api';
 import {
   TeamProjectReference,
   WebApiTeam,
 } from 'azure-devops-node-api/interfaces/CoreInterfaces';
 import { UrlReader } from '@backstage/backend-common';
+import { Config } from '@backstage/config';
+import {
+  AzureDevOpsCredentialsProvider,
+  DefaultAzureDevOpsCredentialsProvider,
+  ScmIntegrations,
+} from '@backstage/integration';
+import {
+  mappedBuildRun,
+  mappedGitTag,
+  mappedPullRequest,
+  mappedRepoBuild,
+} from './mappers';
 
 /** @public */
 export class AzureDevOpsApi {
-  public constructor(
-    private readonly logger: Logger,
-    private readonly webApi: WebApi,
-    private readonly urlReader: UrlReader,
-  ) {}
+  private readonly logger: Logger;
+  private readonly urlReader: UrlReader;
+  private readonly config: Config;
+  private readonly credentialsProvider: AzureDevOpsCredentialsProvider;
+
+  private constructor(
+    logger: Logger,
+    urlReader: UrlReader,
+    config: Config,
+    credentialsProvider: AzureDevOpsCredentialsProvider,
+  ) {
+    this.logger = logger;
+    this.urlReader = urlReader;
+    this.config = config;
+    this.credentialsProvider = credentialsProvider;
+  }
+
+  static fromConfig(
+    config: Config,
+    options: { logger: Logger; urlReader: UrlReader },
+  ) {
+    const scmIntegrations = ScmIntegrations.fromConfig(config);
+    const credentialsProvider =
+      DefaultAzureDevOpsCredentialsProvider.fromIntegrations(scmIntegrations);
+    return new AzureDevOpsApi(
+      options.logger,
+      options.urlReader,
+      config,
+      credentialsProvider,
+    );
+  }
+
+  private async getWebApi(host?: string, org?: string): Promise<WebApi> {
+    // If no host or org is provided we fall back to the values from the `azureDevOps` config section
+    // these may have been setup in the `integrations.azure` config section
+    // which is why use them here and not just falling back on them entirely
+    const validHost = host ?? this.config.getString('azureDevOps.host');
+    const validOrg = org ?? this.config.getString('azureDevOps.organization');
+    const url = `https://${validHost}/${encodeURI(validOrg)}`;
+
+    const credentials = await this.credentialsProvider.getCredentials({
+      url,
+    });
+
+    let authHandler;
+    if (!credentials) {
+      // No credentials found for the provided host and org in the `integrations.azure` config section
+      // use the fall back personal access token from `azureDevOps.token`
+      const token = this.config.getString('azureDevOps.token');
```

Comment:
Alternatively, would a `codemod` be an option to move the `azureDevOps.token` token value to `integrations.azure.credentials[i].personalAccessToken`?

Reply:
My thinking on the final state of this once all the changes are in place is that if we can't find the information we need from the `integration.azure` config then we should throw an error with details as to how to correct it. There is simply no reasonable way we can provide a sound default across an unknown number of Orgs and/or Cloud vs Server setups. 

Now when it comes to the code in flight I'm trying to straddle these two worlds. Supporting what we have currently while also supporting the new `AzureDevOpsCredentialsProvider` so that we can migrate away from the `azureDevOps` config, which only ever supported PATs.

Ideally by the time this all lands Adopters would have already migrated to the new config introduced with the `AzureDevOpsCredentialsProvider` nearly 3 months ago. Given that I'm not sure there's going to be a lot of benefit of creating a `codemod` for this.

@sanderaernouts does that make sense?

Reply:
@awanlin makes sense 👍